pkg/v2/tkr/webhook/cluster/tkr-resolver/cluster/cluster.go

```diff
@@ -6,16 +6,335 @@ package cluster
+	}
+
+	if err := cw.ResolveAndSetMetadata(cluster, clusterClass); err != nil {
+		return admission.Denied(err.Error())
+	}
+
+	return success(&req, cluster)
+}
+
+// ResolveAndSetMetadata uses cw.TKRResolver and injects resolved metadata into the provided cluster.
+// Pre-reqs: cluster != nil && clusterClass != nil
+func (cw *Webhook) ResolveAndSetMetadata(cluster *clusterv1.Cluster, clusterClass *clusterv1.ClusterClass) error {
+	query, err := cw.constructQuery(cluster, clusterClass)
+	if err != nil {
+		return err
+	}
+	if isEmpty(query) { // skipping resolution entirely
+		return nil
+	}
+
+	result := cw.TKRResolver.Resolve(query)
+
+	isUnresolvedCP := isUnresolved(result.ControlPlane)
+	unresolvedMDs := unresolvedMachineDeployments(result)
+	if isUnresolvedCP || len(unresolvedMDs) != 0 {
+		return &errUnresolved{
+			query:   query,
+			result:  result,
+			cluster: cluster,
+			cp:      isUnresolvedCP,
+			mds:     unresolvedMDs,
+		}
+	}
+
+	setMetadata(result, cluster)
+	cw.adjustKubernetesVersion(cluster)
+	return nil
+}
+
+// constructQuery creates TKR resolution query from cluster and clusterClass metadata.
+// Pre-reqs: cluster != nil && clusterClass != nil
+func (cw *Webhook) constructQuery(cluster *clusterv1.Cluster, clusterClass *clusterv1.ClusterClass) (data.Query, error) {
+	tkrSelector, err := selectorFromAnnotation(cluster.Annotations, clusterClass.Annotations, runv1.AnnotationResolveTKR)
+	if tkrSelector == nil || cluster.Spec.Topology == nil {
+		return data.Query{}, err
+	}
+
+	osImageSelector, err := selectorFromAnnotation(
+		cluster.Spec.Topology.ControlPlane.Metadata.Annotations,
+		clusterClass.Spec.ControlPlane.Metadata.Annotations,
+		runv1.AnnotationResolveOSImage)
+	if err != nil {
+		return data.Query{}, err
+	}
+	if osImageSelector == nil {
+		osImageSelector = labels.Everything() // default to empty selector (matches all) for OSImages
+	}
+
+	cpQuery := cw.constructOSImageQuery(cluster.Spec.Topology.Version, tkrSelector, osImageSelector, labels.Merge(cluster.Labels, cluster.Spec.Topology.ControlPlane.Metadata.Labels))
+
+	if cluster.Spec.Topology.Workers == nil {
+		return data.Query{ControlPlane: cpQuery}, nil
+	}
+
+	mdQueries, err := cw.constructMDQueries(cluster, clusterClass, tkrSelector)
+	if err != nil {
+		return data.Query{}, err
+	}
+
+	return data.Query{ControlPlane: cpQuery, MachineDeployments: mdQueries}, nil
+}
+
+// selectorFromAnnotation produces a selector from the value of the specified annotation.
+func selectorFromAnnotation(cAnnots, ccAnnots map[string]string, annotation string) (labels.Selector, error) {
+	var selectorStr *string
+	if selectorStr = getAnnotation(cAnnots, annotation); selectorStr == nil {
+		if selectorStr = getAnnotation(ccAnnots, annotation); selectorStr == nil {
+			return nil, nil
+		}
+	}
+	selector, err := labels.Parse(*selectorStr)
+	return selector, errors.Wrapf(err, "error parsing selector: '%s'", *selectorStr)
```

Comment:
When I expect to see a simple name or something I know is a compact string representation, I always use `'%s'` so that logs would show `''` if an empty string ends up to be the value. 

I don't want to use `%q` because the logger may escape the quotes and it would then look like `\"\"` which I don't like.